### DISCUSSION TITLE
Updating a cached entity by its key without going through fields of Query

### DISCUSSION THREAD

**Discussion by Eliasvdb** [urql-graphql/urql#2845] [Q&A] [answered]
We are currently having some troubles figuring out how to correctly update an entity in the graphcache after a mutation. Our setup is as follows (simplified code for illustration):

We execute a lot of Page queries, they contain components, these can be of different types, on of the types is `PaginatedList`. These lists are normalized by their ID (we also use the `relayPagination()` resolver for this type).
```gql
{
  page {
    id
    components {
      ... on PaginatedList {
        id
        paginatedItems {
          # your usual edges and connections in a relay paginated list
        }
        pageInfo {
          # your usual page info
        }
      }
    }
  }
}
```

That same type is also sometimes fetched directly using a query like (for example to fetch an additional page of a list we got from the previous page query):
```gql
{
  list {
    ... on PaginatedList {
      id
      paginatedItems {
        # your usual edges and connections in a relay paginated list
      }
      pageInfo {
        # your usual page info
      }
    }
  }
}
```

We have a mutation which will remove an item from a list. We want this deletion to be reflected in our cached list. We cannot come up with a way to specifically update a `PaginatedList` entity by its ID, without having to go through the fields of query. In our eyes, this should not be necessary since the type `PaginatedList` is normalized, there is alsways only 1 instance for each unique list (by ID), so why can we not update the entity by its ID? We should be able to fetch this entity on its own and update it, so it is reflected in all parents which have a link to this list.

`writeFragment` does not allow us to remove an item from `paginatedItems` and `updateQuery` requires us to do the action twice, once for page queries:
```js
cache
.inspectFields('Query')
.filter((field) => field.fieldName === 'page')
```
and once for list queries:
```js
cache
.inspectFields('Query')
.filter((field) => field.fieldName === 'list')
```

We can get the ID of the normalized list as follows:
```js
const cacheKey = cache.keyOfEntity({ __typename: 'PaginatedList', id: listId});
```
But we cannot find a way to update the value of its field `paginatedItems`.

What are we doing wrong? Any help would be much appreciated 🙏

**Comment by JoviDeCroock** [comment, marked as answer]
You can do `cache.invalidate({ __typename: 'your-type', id });` this will lead to the list being refetched the next time you are displaying it. If you need a synchronous update `updateQuery` is still your best bet

**Reply by Eliasvdb** [reply]
The problem with invalidating is that it might take a few seconds/ a bit longer for our backend to be updated according to the mutation we did, that's why we want to remove the item from the cache manually, to reflect the change that will happen in de backend but might not have happened yet due to processing time.

**Reply by JoviDeCroock** [reply]
In that case `updateQuery` is your best bet, it will be more code as it kind of goes against the premise that your normalized cache is reflecting the state of your back-end 😅 you are "mirroring" what should have happened on your back-end. It might be worth looking at `cache.link` instead as well though, that could be easier, while going through your lists you can check what the field is linked to and then remove that and link again with `cache.link` https://formidable.com/open-source/urql/docs/api/graphcache/#link

**Reply by Eliasvdb** [reply]
With `updateQuery` our updater looks really complex for the pretty 'simple' use case we are trying to solve here. It feels unnecessary to have to go through all fields of Query which might include our `PaginatedList` component. If in the future another field of Query would potentially contain a `PaginatedList`, we would need to add this field manually even though the list itself (which exists as a separate normalized cache entity) stays the same as before.

**Comment by kitten** [comment]
If I'd have to guess you're probably looking for this section of the docs and for `cache.resolve` and `cache.link` to update just the list field https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#writing-links-individually

**Comment by Eliasvdb** [comment]
Okay the pointers to `cache.resolve` and `cache.link` have helped a lot, thanks @JoviDeCroock and @kitten . How far I've come so far:
```js
const cacheKey = cache.keyOfEntity({ __typename: 'PaginatedList', id: listId});
const paginatedItems = cache.inspectFields(cacheKey).filter(field => field.fieldName === 'paginatedItems');
paginatedItems.forEach(field => {
  const paginatedItemsLink = cache.resolve({ __typename: 'PaginatedList', id: listId}, field.fieldKey);
  const edgeLinks = cache.resolve(paginatedItemsLink, 'edges');
  // this is an array of keys ['PaginatedList:id1.paginatedItems({"first":10}).edges.0, 'PaginatedList:id1.paginatedItems({"first":10}).edges.1, ...]
  const updatedEdges = edgeLinks.filter(edge => cache.resolve(edge, 'cursor') !== args.input.id);
  // imagine we want to remove the first item, updatedEdges would be:
  // ['PaginatedList:id1.paginatedItems({"first":10}).edges.1, ...]
  
  cache.link(paginatedItemsLink, 'edges', updatedEdges);
});
```

This seems to be working as expected! Still quite a few `cache.resolve`'s before we get what we want to remove, but already a whole lot better than going through `updateQuery`!

**Reply by JoviDeCroock** [reply]
What version of graph-cache are you using? So when you check `cache.resolve` for the `paginatedItemsLink, 'edges'` you see only the ones that are still meant to be in there? This isn't reflected to a mounted `useQuery`?

**Reply by Eliasvdb** [reply]
Nevermind, there was an issue with the ID's on our side, it is working as expected, thanks!

**Reply by zoontek** [reply]
I'm trying to understand how can this be used to update some edge data (for a few hours now 😅).

```gql
type Query {
  team(teamId: String!): Team!
}

type Team {
  id: String!
  users(after: String, first: Int!): TeamUsersConnection!
}

type TeamUsersConnectionEdge {
  cursor: String!
  node: User!
  role: Role! # role correspond to the user role in this given team
}
```

```ts
// Used to remove an item from a paginated list (from @Eliasvdb)

const unlinkEdge = (
  cache: Cache,
  { entity, fieldName, cursor }: { entity: Entity; fieldName: string; cursor: string },
) => {
  const fields = cache
    .inspectFields(cache.keyOfEntity(entity))
    .filter((field) => field.fieldName === fieldName);

  fields.forEach(({ fieldKey }) => {
    const field = cache.resolve(entity, fieldKey) as Entity;
    const edges = cache.resolve(field, "edges") as Entity[];

    const filtered = edges.filter(
      (edge) => cache.resolve(edge, "cursor") !== cursor,
    );

    cache.link(field, "edges", filtered);
  });
};

// Usage

const cache = cacheExchange<GraphCacheConfig>({
  schema,

  resolvers: {
    Team: {
      users: relayPagination({ mergeMode: "inwards" }),
    },
  },

  updates: {
    Mutation: {
      updateTeamUser: (_result, args, cache) => {},

      removeTeamUser: (_result, args, cache) => {
        unlinkEdge(cache, {
          entity: { __typename: "Team", id: args.teamId },
          fieldName: "users",
          cursor: args.userId,
        });
      },
    },
  },
});
```

I would like to update `TeamUsersConnectionEdge` `role` value on `updateTeamUser` result, but I cannot find a way how.

```ts
const filtered = edges.map((edge) => {
  if (cursor === cache.resolve(edge, "cursor")) {
    // how can I return a modified edge here?
  }

  return edge;
});

cache.link(field, "edges", filtered);
```
